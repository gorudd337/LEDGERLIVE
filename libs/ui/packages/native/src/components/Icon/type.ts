import React from "react";

export type IconProps = { size?: number; color?: string };
export type IconType = React.ComponentType<IconProps>;
export type IconOrElementType = IconType | React.ReactElement;
