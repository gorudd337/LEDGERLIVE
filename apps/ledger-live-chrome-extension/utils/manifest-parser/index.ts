type Manifest = chrome.runtime.ManifestV3;

class ManifestParser {
  static convertManifestToString(manifest: Manifest): string {
    return JSON.stringify(manifest, null, 2);
  }
}

export default ManifestParser;
