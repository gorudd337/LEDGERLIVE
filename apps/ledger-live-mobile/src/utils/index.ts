export * from "./vibrate";
export * from "./urls";
export * from "./constants";
export * from "./vibrate";
