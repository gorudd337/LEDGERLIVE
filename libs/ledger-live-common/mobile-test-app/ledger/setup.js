// @flow
import { setTestFile } from "../engine";

global._JEST_SETUP = (testName: string) => {
  setTestFile(testName);
};
