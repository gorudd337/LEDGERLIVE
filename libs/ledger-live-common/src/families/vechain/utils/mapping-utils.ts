import vip180 from "../contracts/abis/VIP180";
import { Operation } from "@ledgerhq/types-live";
import BigNumber from "bignumber.js";
import { EventLog, TransferLog } from "../api/types";
import { encodeOperationId } from "@ledgerhq/coin-framework/operation";
import { getFees } from "./transaction-utils";

export const mapVetTransfersToOperations = async (
  txs: TransferLog[],
  accountId: string,
  addr: string,
): Promise<Operation[]> => {
  return Promise.all(
    txs.map(async tx => {
      const fees = await getFees(tx.meta.txID);
      return {
        id: encodeOperationId(
          accountId,
          tx.meta.txID,
          tx.recipient === addr.toLowerCase() ? "IN" : "OUT",
        ),
        hash: tx.meta.txID,
        type: tx.recipient === addr.toLowerCase() ? "IN" : "OUT",
        value: new BigNumber(tx.amount),
        fee: new BigNumber(fees),
        senders: [tx.sender],
        recipients: [tx.recipient],
        blockHeight: tx.meta.blockNumber,
        blockHash: tx.meta.blockID,
        accountId,
        date: new Date(tx.meta.blockTimestamp * 1000),
        extra: {},
      };
    }),
  );
};

export const mapTokenTransfersToOperations = async (
  evnts: EventLog[],
  accountId: string,
  addr: string,
): Promise<Operation[]> => {
  return Promise.all(
    evnts.map(async evnt => {
      const decoded = vip180.TransferEvent.decode(evnt.data, evnt.topics);
      const fees = await getFees(evnt.meta.txID);
      return {
        id: encodeOperationId(
          accountId,
          evnt.meta.txID,
          decoded.to === addr.toLowerCase() ? "IN" : "OUT",
        ),
        hash: evnt.meta.txID,
        type: decoded.to === addr.toLowerCase() ? "IN" : "OUT",
        value: new BigNumber(decoded.value),
        fee: fees,
        senders: [decoded.from],
        recipients: [decoded.to],
        blockHeight: evnt.meta.blockNumber,
        blockHash: evnt.meta.blockID,
        accountId,
        date: new Date(evnt.meta.blockTimestamp * 1000),
        extra: {},
      };
    }),
  );
};

/**
 * Scale the number up by the specified number of decimal places
 * @param val - the value to scale up (a number or string representation of a number)
 * @param scaleDecimal - the number of decimals to scale up by
 * @param roundDecimal - the number of decimals to round the result to
 * @param roundingStrategy - what strategy to use when rounding. Based on the strategies defined in `bignumber.js`. Default strategy is ROUND_HALF_UP
 * @returns the scaled up result as a string
 */
export const scaleNumberUp = (
  val: BigNumber.Value,
  scaleDecimal: number,
  roundDecimal = 0,
  roundingStrategy: BigNumber.RoundingMode = BigNumber.ROUND_HALF_UP,
): string => {
  if (scaleDecimal === 0) return new BigNumber(val).toFixed();
  if (scaleDecimal < 0) throw Error("Decimal value must be greater than or equal to 0");
  const valBn = new BigNumber(val);
  if (valBn.isNaN()) throw Error("The value provided is NaN.");

  const amount = valBn.times(`1${"0".repeat(scaleDecimal)}`);

  if (scaleDecimal === roundDecimal) return amount.toFixed();

  return amount.toFixed(roundDecimal, roundingStrategy);
};
