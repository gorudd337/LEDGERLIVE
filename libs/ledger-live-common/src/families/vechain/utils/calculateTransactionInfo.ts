import { Account, TokenAccount } from "@ledgerhq/types-live";
import BigNumber from "bignumber.js";
import { Transaction, TransactionInfo } from "../types";
import { calculateFee } from "../utils/transaction-utils";
import { DEFAULT_GAS_COEFFICIENT } from "../constants";
// import { calculateFee, estimateGas } from "./transaction-utils";
// import { isValid } from "./address-utils";
// import { calculateClausesVtho } from "../js-transaction";

export const calculateTransactionInfo = async (
  account: Account,
  transaction: Transaction,
): Promise<TransactionInfo> => {
  const { subAccounts } = account;
  const { amount: oldAmount, useAllAmount, subAccountId } = transaction;
  const maxTokenFees = await calculateMaxFeesToken();

  const tokenAccount =
    subAccountId && subAccounts
      ? (subAccounts.find(subAccount => {
          return subAccount.id === subAccountId;
        }) as TokenAccount)
      : undefined;
  const isTokenAccount = !!tokenAccount;

  let balance;
  let spendableBalance;
  let amount;

  if (isTokenAccount) {
    balance = tokenAccount.balance;
    spendableBalance = tokenAccount.balance.minus(maxTokenFees).gt(0)
      ? tokenAccount.balance.minus(maxTokenFees)
      : new BigNumber(0);
    amount = useAllAmount ? spendableBalance : oldAmount;
  } else {
    balance = account.balance;
    spendableBalance = account.balance;
    amount = useAllAmount ? spendableBalance : oldAmount;
  }

  return {
    isTokenAccount,
    amount,
    spendableBalance,
    balance,
    tokenAccount,
  };
};

export const calculateMaxFeesToken = async (): Promise<BigNumber> => {
  //FIXME: The BE call is currently returning not stable values, hardwiring a value
  // account: AccountLike,
  // transaction: Transaction

  // const accountTmp =
  //   account.type === "Account" ? account?.subAccounts?.[0] : account;

  // if (
  //   transaction.subAccountId &&
  //   transaction.recipient &&
  //   isValid(transaction.recipient) &&
  //   accountTmp
  // ) {
  //   transaction.amount = new BigNumber("1500000000000000");
  //   const clauses = await calculateClausesVtho(
  //     transaction,
  //     new BigNumber("1500000000000000")
  //   );
  //   console.warn(transaction);
  //   const gas = await estimateGas({
  //     ...transaction,
  //     body: { ...transaction.body, clauses: clauses },
  //   });
  //   const estimatedFees = new BigNumber(gas);
  //   return estimatedFees;
  // }
  const fees = calculateFee(new BigNumber("67000"), DEFAULT_GAS_COEFFICIENT);
  return fees;
};

export const calculateTotalSpent = (isToken: boolean, transaction: Transaction): BigNumber => {
  if (isToken) {
    return transaction.amount.plus(transaction.estimatedFees);
  } else {
    return transaction.amount;
  }
};
