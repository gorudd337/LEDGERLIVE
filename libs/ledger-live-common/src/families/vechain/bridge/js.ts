import type { Transaction } from "../types";
import { makeAccountBridgeReceive } from "../../../bridge/jsHelpers";

import { sync, scanAccounts } from "../js-synchronisation";
import { AccountBridge, AccountLike, CurrencyBridge } from "@ledgerhq/types-live";
import { createTransaction, updateTransaction, prepareTransaction } from "../js-transaction";
import getTransactionStatus from "../js-getTransactionStatus";
import signOperation from "../js-signOperation";
import broadcast from "../js-broadcast";
import BigNumber from "bignumber.js";
import { calculateMaxFeesToken } from "../utils/calculateTransactionInfo";

const receive: AccountBridge<Transaction>["receive"] = makeAccountBridgeReceive();

const currencyBridge: CurrencyBridge = {
  scanAccounts,
  preload: async (): Promise<Record<string, any>> => {
    return {};
  },
  hydrate: (): void => {},
};

const estimateMaxSpendable = async (inputs: {
  account: AccountLike;
  transaction: Transaction;
}): Promise<BigNumber> => {
  const { account, transaction } = inputs;

  if (account.type === "Account") {
    return account.balance;
  }
  if (transaction) {
    const maxTokenFees = await calculateMaxFeesToken();
    const spendable = account.balance.minus(maxTokenFees);
    if (spendable.gt(0)) return account.balance.minus(maxTokenFees);
    return new BigNumber(0);
  } else {
    return account.balance;
  }
};

const accountBridge: AccountBridge<Transaction> = {
  estimateMaxSpendable,
  createTransaction,
  updateTransaction,
  getTransactionStatus,
  prepareTransaction,
  sync,
  receive,
  signOperation,
  broadcast,
};

export default { currencyBridge, accountBridge };
