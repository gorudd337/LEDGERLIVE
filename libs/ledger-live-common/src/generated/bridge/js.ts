import bitcoin from "../../families/bitcoin/bridge/js";
import cardano from "../../families/cardano/bridge/js";
import celo from "../../families/celo/bridge/js";
import cosmos from "../../families/cosmos/bridge/js";
import crypto_org from "../../families/crypto_org/bridge/js";
import elrond from "../../families/elrond/bridge/js";
import ethereum from "../../families/ethereum/bridge/js";
import filecoin from "../../families/filecoin/bridge/js";
import hedera from "../../families/hedera/bridge/js";
import internet_computer from "../../families/internet_computer/bridge/js";
import near from "../../families/near/bridge/js";
import ripple from "../../families/ripple/bridge/js";
import solana from "../../families/solana/bridge/js";
import stacks from "../../families/stacks/bridge/js";
import stellar from "../../families/stellar/bridge/js";
import tezos from "../../families/tezos/bridge/js";
import tron from "../../families/tron/bridge/js";
import vechain from "../../families/vechain/bridge/js";
import { makeLRUCache } from "@ledgerhq/live-network/cache";
import network from "@ledgerhq/live-network/network";
import { withDevice } from "../../hw/deviceAccess";
import { createBridges as polkadotCreateBridges } from "@ledgerhq/coin-polkadot/bridge/js";
import { createBridges as algorandCreateBridges } from "@ledgerhq/coin-algorand/bridge/js";
import { createBridges as evmCreateBridges } from "@ledgerhq/coin-evm/bridge/js";

export default {
  bitcoin,
  cardano,
  celo,
  cosmos,
  crypto_org,
  elrond,
  ethereum,
  filecoin,
  hedera,
  internet_computer,
  near,
  ripple,
  solana,
  stacks,
  stellar,
  tezos,
  tron,
  vechain,
  algorand,
  evm,
  polkadot,
};
