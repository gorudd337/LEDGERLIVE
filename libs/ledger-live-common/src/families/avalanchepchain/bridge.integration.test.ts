import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { DatasetTest, CurrenciesData } from "@ledgerhq/types-live";
import type { AvalanchePChainAccountRaw, Transaction } from "./types";
import { BigNumber } from "bignumber.js";
import {
  InvalidAddressBecauseDestinationIsAlsoSource,
  AmountRequired,
} from "@ledgerhq/errors";
import { FIVE_MINUTES, TWO_WEEKS, AVAX_MINIMUM_STAKE_AMOUNT } from "./utils";
import {
  AvalancheInvalidDateTimeError,
  AvalancheMinimumAmountError,
} from "./errors";

const avalanchepchain: CurrenciesData<Transaction, AvalanchePChainAccountRaw> =
  {
    scanAccounts: [
      {
        name: "avalanchepchain seed 1",
        apdus: `
      => 800300000f0000038000002c8000232880000000
      <= 2103828c8dad550959fba4b8b7aef4fff624ff514acbd5e5ea8095b2e714747eed7e0572f641ad2ae06b7d0d0e190e74af86732927ced7b828fef23589257bbed1e59000
      `,
      },
    ],
    FIXME_ignorePreloadFields: ["validators"],
    accounts: [
      {
        FIXME_tests: [
          "balance is sum of ops",
          "invalid recipient have a recipientError",
        ],
        raw: {
          id: "js:1:avalanchepchain:avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5:",
          seedIdentifier: "",
          xpub: "xpub6Cfxnaft2T7TZ5sr3istgn2uvjTzunaZdL1rGpNU3orLUQggKYFxHSo3Csf98QwJA7BASwmGWNTwu1tc7MMRh3TvkJ6aDV55f7re6wnj42t",
          derivationMode: "",
          index: 0,
          freshAddress: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
          freshAddressPath: "44'/9000'/0'/0/0",
          freshAddresses: [
            {
              address: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
              derivationPath: "44'/9000'/0'/0/0",
            },
          ],
          name: "Avalanche P-chain 1 - Nano X",
          balance: "0",
          blockHeight: 2168241,
          currencyId: "avalanchepchain",
          unitMagnitude: 9,
          operationsCount: 0,
          operations: [],
          pendingOperations: [],
          lastSyncDate: "",
          avalanchePChainResources: {
            publicKey:
              "03828c8dad550959fba4b8b7aef4fff624ff514acbd5e5ea8095b2e714747eed7e",
            chainCode:
              "0572f641ad2ae06b7d0d0e190e74af86732927ced7b828fef23589257bbed1e5",
            stakedBalance: "",
            delegations: [],
          },
        },
        transactions: [
          {
            name: "Same as Recipient",
            transaction: (t) => ({
              ...t,
              amount: new BigNumber(100),
              recipient: "avax10u6rtayzscc8ds4l5tdyjm2uazh0enhj8gguh5",
            }),
            expectedStatus: {
              errors: {
                recipient: new InvalidAddressBecauseDestinationIsAlsoSource(),
              },
              warnings: {},
            },
          },
          {
            name: "Delegation - Not enough balance to delegate",
            transaction: (t) => ({
              ...t,
              mode: "delegate",
              amount: new BigNumber(100),
            }),
            expectedStatus: {
              errors: {
                amount: new AvalancheMinimumAmountError(),
              },
              warnings: {},
            },
          },
          {
            name: "Delegation - AmountRequired",
            transaction: (t) => ({
              ...t,
              mode: "delegate",
              amount: new BigNumber(0),
            }),
            expectedStatus: {
              errors: {
                amount: new AmountRequired(),
              },
              warnings: {},
            },
          },
          {
            name: "Delegation - Under time threshold",
            transaction: (t) => ({
              ...t,
              mode: "delegate",
              amount: new BigNumber(AVAX_MINIMUM_STAKE_AMOUNT),
              recipient: "GW7CvXwkPFDsfFHrix1SA148NHF5ZnSXs",
              endTime: new BigNumber(
                Math.round(new Date().getTime() / 1000) + TWO_WEEKS
              ),
            }),
            expectedStatus: {
              errors: {
                time: new AvalancheInvalidDateTimeError(),
              },
              warnings: {},
            },
          },
          {
            name: "Delegation - Beyond time threshold",
            transaction: (t) => ({
              ...t,
              mode: "delegate",
              amount: new BigNumber(AVAX_MINIMUM_STAKE_AMOUNT),
              recipient: "GW7CvXwkPFDsfFHrix1SA148NHF5ZnSXs",
              endTime: new BigNumber(
                Math.round(new Date().getTime() / 1000) +
                  TWO_WEEKS +
                  FIVE_MINUTES
              ),
              maxEndTime: new BigNumber(
                Math.round(new Date().getTime() / 1000) + TWO_WEEKS
              ),
            }),
            expectedStatus: {
              errors: {
                time: new AvalancheInvalidDateTimeError(),
              },
              warnings: {},
            },
          },
        ],
      },
    ],
  };

const dataset: DatasetTest<Transaction> = {
  implementations: ["js"],
  currencies: {
    avalanchepchain,
  },
};

testBridge(dataset);
