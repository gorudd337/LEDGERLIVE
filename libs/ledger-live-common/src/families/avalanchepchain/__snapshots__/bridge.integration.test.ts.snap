// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avalanchepchain currency bridge scanAccounts avalanchepchain seed 1 1`] = `
Array [
  Object {
    "avalanchePChainResources": Object {
      "chainCode": "0572f641ad2ae06b7d0d0e190e74af86732927ced7b828fef23589257bbed1e5",
      "delegations": Array [],
      "publicKey": "03828c8dad550959fba4b8b7aef4fff624ff514acbd5e5ea8095b2e714747eed7e",
      "stakedBalance": "0",
    },
    "balance": "20000000",
    "currencyId": "avalanchepchain",
    "derivationMode": "avalanchepchain",
    "freshAddress": "",
    "freshAddressPath": "m/44'/9000'/0'",
    "freshAddresses": Array [
      Object {
        "address": "",
        "derivationPath": "m/44'/9000'/0'",
      },
    ],
    "id": "js:2:avalanchepchain::avalanchepchain",
    "index": 0,
    "name": "Avalanche P-Chain 1",
    "nfts": undefined,
    "operationsCount": 1,
    "pendingOperations": Array [],
    "seedIdentifier": "03828c8dad550959fba4b8b7aef4fff624ff514acbd5e5ea8095b2e714747eed7e",
    "spendableBalance": "20000000",
    "starred": false,
    "swapHistory": Array [],
    "syncHash": undefined,
    "unitMagnitude": 9,
    "used": true,
  },
]
`;

exports[`avalanchepchain currency bridge scanAccounts avalanchepchain seed 1 2`] = `
Array [
  Array [
    Object {
      "accountId": "js:2:avalanchepchain::avalanchepchain",
      "blockHash": "2iMUqvW5t3mBBvy2YjnJtg5ffuRY4bWyhSktJPu2oFi1tDdYAt",
      "blockHeight": 4279835,
      "contract": undefined,
      "extra": Object {},
      "fee": "1000000",
      "hash": "23yMrAvxeHoPFivFunG35i8iXEqXoYgLHeV4f8CV33JqF3GXFh",
      "id": "js:2:avalanchepchain::avalanchepchain-23yMrAvxeHoPFivFunG35i8iXEqXoYgLHeV4f8CV33JqF3GXFh-IN",
      "operator": undefined,
      "recipients": Array [
        "avax13q2k264xx9l8lf7qeayun8cjle5p9j35pjzmgp",
      ],
      "senders": Array [
        "avax13q2k264xx9l8lf7qeayun8cjle5p9j35pjzmgp",
      ],
      "standard": undefined,
      "tokenId": undefined,
      "type": "IN",
      "value": "20000000",
    },
  ],
]
`;
